### ISSUE TITLE
Looking for a plugin hook that runs early enough to change which specs Cypress will run

### ISSUE THREAD

**Issue by istateside** [cypress-io/cypress#25404]
### What would you like?

Allow plugins to change which specs will be run by Cypress

### Why is this needed?

I would like to use the plugin API to change which spec files are run by Cypress. I originally thought this would've been doable with `before:spec` or `before:run`, but neither seems to work for what I need. There doesn't appear to be a way to "skip" a test from the `before:spec` hook, and trying to manipulate the `details` object provided in the `before:run` hook does not seem to affect which tests Cypress runs.

I assumed that I could change the `details.specs` array and remove specs that I didn't want to be run, but that did not affect the tests that were included in the subsequent Cypress run. I also tried modifying the `excludeSpecPattern` value in `details.config`, but it looks like the included specs are already determined by that point.

Ideally, we could use `before:run` to do this, but maybe a hook for a new point in the plugin lifecycle would be needed.

Assuming the `details` object was something close what we get in the `before:run` hook, the idea is something like this:

```js
on('before:some-event', async (details) => {
  const dataFromSomeAPI = await getTestDataFromAPI(); // some network request
  const badTestNames = dataFromSomeAPI.namesOfTestsWeDontLike

  return {
    ...details,
    specs: details.specs.filter(s => !badTestNames.includes(s.name));
  };
}))
```

**Comment by emilyrohrbough** [comment]
@istateside Its not completely clear why you are wanting to filter specs after Cypress has started. Can you share a bit more on why you are trying to do this?

Also, have you try leveraging the `setupNodeEvents` in Cypress 10+? Any configuration updates returned here will be used in the run:
```js
module.exports = defineConfig({
  // setupNodeEvents can be defined in either
  // the e2e or component configuration
  e2e: {
    setupNodeEvents(on, config) {
     const dataFromSomeAPI = await getTestDataFromAPI(); // some network request
     const badTestNames = dataFromSomeAPI.namesOfTestsWeDontLike
 
     config. excludeSpecPattern = dataFromSomeAPI.namesOfTestsWeDontLike

      // return the updated config object
      return config
    }
  }
})
```

If you are looking specifically to skip specific tests in a spec file, you can also perform and in-line skip based on a specific set of criteria.
```js
describe('suite', () => {
  it('t1', function () {
      if (shouldBeSkipped) {
         this.skip()
      }
  })
})
```

**Comment by istateside** [comment]
@emilyrohrbough Thanks for the reply - I had originally typed this up with more context around our specific usecase, but ended up trimming the report down to be agnostic of our setup.

Looking at your example again, I realized I had made an error in my attempt at getting this set up - I was trying to return an updated configuration object inside of a function running as the `before:run` hook. I didn't realize that I needed to return the object inside the `setupNodeEvents` hook, one level higher.

So I was doing this:
```js
setupNodeEvents(on, config) {
  on('before:run', async details => {
    const badTests = await getBadTestNames();
    const newExcludeSpecPattern = getExcludeSpec(badTests);
    return { excludeSpecPattern: newExcludeSpecPattern }; // This does nothing in before:run
  });
}
```

instead of this:

```js
setupNodeEvents(on, config) {
  const badTests = await getBadTestNames();
  const newExcludeSpecPattern = getExcludeSpec(badTests);
  return { excludeSpecPattern: newExcludeSpecPattern }; // This works as intended
}
```

Returning a config object like you shared works exactly how I want it to - thank you!

> Its not completely clear why you are wanting to filter specs after Cypress has started. Can you share a bit more on why you are trying to do this?

My problem is solved, but I'll share our usecase anyway. My team has our own version of a feature that was at one time on the Cypress Dashboard roadmap, which I believe you called "Test Muting" - we call it "quarantining."

We want a tool that fetches a list of tests which we have marked as "quarantined" in an external service (the `getTestDataFromAPI()` bit in my code snippet). We want Cypress to skip any tests that are "quarantined." 

In Cypress <= 9, we had a wrapper library around Cypress that expected your configuration to be JSON - we would read the configuration, and search for all of the test files that should be run based on the config. Then we'd get the list of bad tests, and remove those tests from the list. We'd force Cypress to skip running the bad tests by explicitly setting a `$SPEC` env variable before calling `cypress.run()`

Cypress 10 dropped JSON configuration support, so we needed to either update our wrapper tool to read the config a new way, or find a new way to filter out tests. I wanted to drop the wrapper entirely anyway, so I was looking for ways to do this inside of the Cypress configuration, so users could run `cypress run` directly, without expecting everyone to know to use our own wrapper around Cypress.
